delta/plugins/elasticsearch/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/elasticsearch/indexing/ElasticSearchIndexingCoordinator.scala

```diff
@@ -0,0 +1,157 @@
+import monix.bio.{IO, Task}
+import monix.execution.Scheduler
+import org.apache.jena.rdf.model.Property
+
+private class IndexingStream(
+    client: ElasticSearchClient,
+    viewRes: IndexingViewResource,
+    config: ElasticSearchViewsConfig
+)(implicit cr: RemoteContextResolution, baseUri: BaseUri) {
+  private val view: IndexingElasticSearchView         = viewRes.value
+  private val index: IndexLabel                       = IndexLabel.fromView(config.indexing.prefix, view.uuid, viewRes.rev)
+  private val ctx: ContextValue                       = ContextValue(contexts.elasticSearchIndexing)
+  private val originalSource: Property                = predicate(nxv.originalSource.iri)
+  implicit private val projectionId: ViewProjectionId = viewRes.projectionId
+
+  private def deleteOrIndex(res: ResourceF[IndexingData]): Task[ElasticSearchBulk] =
+    if (res.deprecated && !view.includeDeprecated) delete(res)
+    else index(res)
+
+  private def delete(res: ResourceF[IndexingData]): Task[ElasticSearchBulk] =
+    Task.pure(ElasticSearchBulk.Delete(index, res.id.toString))
+
+  private def index(res: ResourceF[IndexingData]): Task[ElasticSearchBulk] =
+    toDocument(res).map(doc => ElasticSearchBulk.Index(index, res.id.toString, doc))
+
+  private def toDocument(res: ResourceF[IndexingData]): Task[Json] = {
+    val g = res.value.selectPredicatesGraph
+    (if (view.includeMetadata) res.void.toGraph.mapError(illegalArgument).map(_ ++ g) else Task.pure(g)).flatMap {
+      case graph if view.sourceAsText =>
+        val jsonLd = graph.add(originalSource, obj(res.value.source.noSpaces)).toCompactedJsonLd(ctx)
+        jsonLd.mapError(illegalArgument).map(_.json.removeKeys(keywords.context))
+      case graph                      =>
+        val jsonLd = graph.toCompactedJsonLd(ctx)
+        jsonLd.mapError(illegalArgument).map(ld => res.value.source deepMerge ld.json.removeKeys(keywords.context))
+    }
+  }
+
+  private def illegalArgument[A](error: A) =
+    new IllegalArgumentException(error.toString)
+
+  private def containsSchema[A](res: ResourceF[A]): Boolean =
+    view.resourceSchemas.isEmpty || view.resourceSchemas.contains(res.schema.iri)
+
+  private def containsTypes[A](res: ResourceF[A]): Boolean =
+    view.resourceTypes.isEmpty || view.resourceTypes.intersect(res.types).nonEmpty
+
+  def build(
+      eventLog: GlobalEventLog[Message[ResourceF[IndexingData]]],
+      projection: Projection[Unit],
+      initialProgress: ProjectionProgress[Unit]
+  )(implicit sc: Scheduler): IO[Nothing, Stream[Task, Unit]] =
+    for {
+      _     <- client.createIndex(index, Some(view.mapping), view.settings).hideErrorsWith(illegalArgument)
+      eLog  <- eventLog.stream(view.project, initialProgress.offset, view.resourceTag).hideErrorsWith(illegalArgument)
+      stream = eLog
+                 .evalMapFilterValue {
+                   case res if containsSchema(res) && containsTypes(res) => deleteOrIndex(res).map(Some.apply)
```

Comment:
We could try to pass these predicates  to the eventLog, so as to filter out the earliest possible ?

Reply:
That does not seem intuitive

Reply:
We already pass some view information to filter by tag, we could add these two to avoid filtering at several places during the indexing and to avoid unnecessary operations

Reply:
I don't think that is easier to what we are currently doing. Nor more efficient.  So I will discard this change for now.